multi-layer-perceptron/jax/mlp.py

```diff
@@ -0,0 +1,188 @@
+import time
+import numpy as np
+import jax.numpy as jnp
+from tqdm import tqdm
+from jax import grad, jit, vmap
+from jax import random
+from jax.scipy.special import logsumexp
+
+class MLPModel:
+
+    def __init__(self, layer_sizes, num_epochs=1, batch_size=128, n_targets=10, learning_rate=0.01):
+
+        # Array of each layer size
+        self.layer_sizes = layer_sizes
+
+        # Also refered to as the step size
+        # Scaling apply to estimated weight error during backpropagation
+        self.learning_rate = learning_rate
+
+        # Configure parameters
+        self.num_epochs = num_epochs
+        self.batch_size = batch_size
+        self.n_targets = n_targets
+
+        # Initialize all layers with random values
+        self.params = self.__init_layers()
+
+    def __init_layer_params(self, m, n, key, scale=1e-2):
+        """
+        A helper function to randomly initialize weights and biases
+        for a dense neural network layer.
+        """
+        w_key, b_key = random.split(key)
+        
+        return scale * random.normal(w_key, (n, m)), scale * random.normal(b_key, (n,))
+
+    def __init_layers(self, key=random.PRNGKey(0) ):
+        """
+        Initialize all layers for a fully-connected neural network.
+        """
+        sizes = self.layer_sizes
+        keys = random.split(key, len(sizes))
+
+        return [self.__init_layer_params(m, n, k) for m, n, k in zip(sizes[:-1], sizes[1:], keys)]
+
+def relu(x):
+    """
+    Rectified Linear Operation.
+    """
+    return jnp.maximum(0, x)
+
+def one_hot_encoder(x, k, dtype=jnp.float32):
+    """
+    Create a one-hot encoding of x of size k.
+    """
+    return jnp.array(x[:, None] == jnp.arange(k), dtype)
+
+def predict(params, input):
+    """
+    Run a single prediction on the input. 
+
+    Parameters: 
+        input (array): A n-dimensional array, typically a 1D or 2D array. For images for instance this is a 2D array. 
+    """
+
+    activations = input
+    for w, b in params[:-1]:
+        outputs = jnp.dot(w, activations) + b
+        activations = relu(outputs)
+    
+    final_w, final_b = params[-1]
+    logits = jnp.dot(final_w, activations) + final_b
+    
+    return logits - logsumexp(logits)
+
+# Temp, move this
+batched_predict = vmap(predict, in_axes=(None, 0))
+
+def loss(params, inputs, targets):
+    """
+    Compute loss
+    """
+    preds = batched_predict(params, inputs)
+    loss = -jnp.mean(preds * targets)
+
+    return loss
+
+@jit
+def update(params, x, y, learning_rate):
```

Comment:
This is simple stochastic gradient descent, not Adam. It's kind of a pain to implement Adam from scratch (although doable). Might want to use an existing optimizer in Flax or Haiku.

Reply:
have yall looked into optax?

Reply:
Yeah, looking into it I think I have seomthing I can use to implement adam relitively quickly. I think I'll do that first, then I might look into what it would take to implement adam from scratch as it might be a good exercise I think.